tools/js2c.py

```diff
@@ -119,17 +119,19 @@ def GetDefinition(var, source, step=30):
   return definition, len(code_points)
 
 
-def AddModule(filename, definitions, initializers):
+def AddModule(root, filename, definitions, initializers):
   code = ReadFile(filename)
-  name = NormalizeFileName(filename)
+  name = NormalizeFileName(root, filename)
   slug = SLUGGER_RE.sub('_', name)
   var = slug + '_raw'
   definition, size = GetDefinition(var, code)
   initializer = INITIALIZER.format(name, var, size)
   definitions.append(definition)
   initializers.append(initializer)
 
-def NormalizeFileName(filename):
+def NormalizeFileName(root, filename):
+  if root:
+    filename = os.path.relpath(filename, root)
```

Comment:
This doesn't look like the right place to do it - I think the root should be used when searching and reading files, not when normalizing the file names. What is the command supposed to look like with `--root`? Is `SearchFiles` supposed to return files with the root name somwhow?

Reply:
In current build configurations (`gyp/makefile` and `gyp/msvc`), `js2c.py` is executed like this:

```bash
# From source root dir:
tools/js2c.py --target=$target_gen_dir/node_javascript.cc \
              --directory=lib \
              deps/v8/tools/splaytree.mjs deps/v8/tools/codemap.mjs ...
```

From `gyp/ninja` or other build systems like `GN` and `cmake`, it should be executed like this:

```bash
# From build dir like out/Release:
../../tools/js2c.py --target=$target_gen_dir/node_javascript.cc \
                    --directory=../../lib \
                    --root=../.. \
                    ../../deps/v8/tools/splaytree.mjs ../../deps/v8/tools/codemap.mjs ...
```

or this when build dir is on a different partition:

```bash
# From build dir like out/Release:
/path/to/node/tools/js2c.py --target=$target_gen_dir/node_javascript.cc \
                            --directory=/path/to/node/lib \
                            --root=/path/to/node/ \
                            /path/to/node/deps/v8/tools/splaytree.mjs /path/to/node/deps/v8/tools/codemap.mjs ...
```

There are a few things to notice:

1) The `node_library_files` (like `lib/module.js`) are passed via `--directory=lib` with a `SearchFiles`, instead of passing filenames directly like the `deps_files` (the `deps/v8/tools/splaytree.mjs` files). This is to work around windows cmd length limitation (https://github.com/nodejs/node/pull/39069), but not really needed in `gyp/ninja` and other modern build systems, because such commands are executed via [rsp files](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-response-files?view=vs-2022).

2) While executing `js2c.py` in source root dir seems to look more nature, it was based on the idea that each project only has only one build dir, and it usually has side effect of polluting source root dir with build files. Most modern build systems choose to have multiple build dirs (like `out/Debug` and `out/Release`) and contain all build files inside those dirs, and execute scrips inside the build dirs.

So `SearchFiles` does not really matter here, because it does not handle all files, and it is a workaround that should eventually be removed (node should really at least migrate to `gyp/ninja` to avoid handling all the weird things of msvc). And I think `NormalizeFileName` is the proper place to do this, since its job, is to convert the filename to module name.

Reply:
Thanks for the explanation - I don't think `SearchFiles` should be going away though, personally I do not miss the days when we had to modify node.gyp whenever a new JS file is added, it's much nicer that js2c just handles them automatically. It seems simpler if we just do

```
/path/to/node/tools/js2c.py --target=$target_gen_dir/node_javascript.cc \
                            --directory=lib \
                            --root=/path/to/node/ \
                            deps/v8/tools/splaytree.mjs \
                            deps/v8/tools/codemap.mjs ...
```

And maintain that only the `--target` file and the `--root` are actual paths, whereas `--directory` and the rest are paths relative to `--root`. Then specifying `--root` has the same effect of switching to the build directory before running the command. We have to require that `--directory` and the file names must be relative to `--root` anyway, otherwise this could be a bit awkward:

```
/path/to/node/tools/js2c.py --target=$target_gen_dir/node_javascript.cc \
                            --directory=lib \
                            --root=/workspace/node/ \
                            /workspace/other/deps/v8/tools/splaytree.mjs \
                            /workspace/other/deps/v8/tools/codemap.mjs ...
```

Then we end up with a `../other/v8/tools/splaytree.mjs` after path resolution, which isn't going to work.

Reply:
Also as a head up: I am porting js2c.py to C++ in here, to enable better compression support as well as speeding up the build process a bit: https://github.com/nodejs/node/pull/46997